web/src/components/welcome-modal/welcome-modal.tsx

```diff
@@ -27,24 +32,52 @@ export default ({ challengeToken, teamToken, ...props }: WelcomeModalProps) => {
   const [hasAgreed, setHasAgreed] = useState<boolean>(false);
   const account = useAccount();
   const saveAccount = useAction(User.actions.saveAccount);
+  const [locale, toLocaleRoute] = useLocale();
 
   useEffect(() => trackChallenge('modal-welcome'), []);
 
-  const parseEnrollment = (queryString: string): Enrollment => {
-    const regex = new RegExp(/([a-z]+)=([a-z]+)/, 'gm');
+  const parseEnrollment = (
+    queryString: string,
+    referer?: string
+  ): Enrollment => {
+    const regex = new RegExp(/([\w\-]+)=([\w\-]+)/, 'g');
     const queries = {} as { [key: string]: string };
     let pair: Array<string> = [];
 
     while ((pair = regex.exec(queryString)) !== null) {
       queries[pair[1]] = pair[2];
     }
 
-    return {
-      challenge: queries.challenge,
-      team: queries.team,
-      referer: queries.referer,
-      invite: queries.invite,
-    } as Enrollment;
+    queries.referer = referer;
+
+    if (
+      challengeTokens.includes(queries.challenge as ChallengeToken) &&
+      challengeTeamTokens.includes(queries.team as ChallengeTeamToken)
+    ) {
+      return pick(queries, 'challenge', 'team', 'invite') as Enrollment;
+    } else return null;
+  };
+
+  const redirectEnrollment = (enrollmentDetails: string, referrer?: string) => {
+    const referrerString = referrer ? `&referer=${referrer}` : '';
+
+    if (enrollmentDetails) {
+      if (account) {
+        const enrollObject = parseEnrollment(enrollmentDetails, referrer);
+
+        saveAccount({ enrollment: enrollObject }).then(() => {
+          const redirect = {
+            pathname: toLocaleRoute(URLS.DASHBOARD + '/' + URLS.CHALLENGE),
+            search: `?challenge=${enrollObject.challenge}&achievement=1${referrerString}`,
+          };
+          return <Redirect to="{redirect}" />;
```

Comment:
The quotes here mean we're redirecting to the literal `./{redirect}`.

Also: I think we decided last week to start adding the `push` attribute to these. If you think it's best to not use it in this case I probably agree, just wanted to call it out in case.